### PR TITLE
Task 3.3: extraction orchestration and fallback logic

### PR DESCRIPTION
**Pull request by Frank-Yong**
## Summary
- add extraction orchestration API in `docdown/stages/extract.py`
- implement per-chunk primary/fallback flow with `ExtractionResult` tracking and `ExtractorUsed` enum
- add global GROBID availability gate: when unreachable, route all chunks directly to fallback
- keep chunk-level failure isolation so failed chunks do not block others
- add extraction summary logging (grobid/pdfminer/failed counts)
- add orchestration unit tests for all-succeed, partial-failure, full-grobid-down, and all-fail scenarios
- update Task 3.3 checklist and artifact diagram

## Testing
- `pytest tests/test_extract.py -q`
- `pytest -q`

### CODE REVIEW

File: docdown/stages/extract.py [hunk 1]
```diff
@@ -201,6 +220,114 @@ def extract_pdfminer_chunk(
...
+                "GROBID unavailable; using fallback extractor '%s' for all chunks: %s",
+                fallback_extractor,
+                exc,
+            )
+
+    results: list[ExtractionResult] = []
+
+    for chunk_path in chunk_paths:
+        chunk = Path(chunk_path)
+        chunk_number = _chunk_number_from_path(chunk)
```
[the remark this answers is not in the document]
Comment (Frank-Yong) [reply docdown/stages/extract.py]: I’ve applied the per-chunk error handling and added a regression test; next I’m running test_extract.py to confirm the fix and no regressions.

Ran terminal command: $pyRoot='C:\Users\email\AppData\Local\Programs\Python\Python314'; $scriptsDir=Join-Path $pyRoot 'Scripts'; $env:PATH = "$pyRoot;$scriptsDir;" + $env:PATH; pytest tests/test_extract.py -q

Good catch from review, fixed.

### What changed
In extract.py, `orchestrate_extraction(...)` now handles invalid chunk filenames per chunk:

- Wrapped `_chunk_number_from_path(chunk)` in `try/except ValueError`.
- On parse failure:
  - logs an error for that chunk,
  - appends a failed `ExtractionResult`,
  - continues to the next chunk instead of aborting orchestration.

This preserves the “failed chunks do not block others” behavior.

### Test added
In test_extract.py:
- `test_orchestrate_extraction_invalid_chunk_name_does_not_block_others`
  - uses `bad-name.pdf` followed by `chunk-0002.pdf`
  - asserts first result is failed with parse error
  - asserts second chunk still succeeds via GROBID path

### Validation
- `pytest test_extract.py -q` -> `24 passed, 1 skipped` (plus existing `RequestsDependencyWarning`)

File: docdown/stages/extract.py [hunk 2]
```diff
@@ -201,6 +220,114 @@ def extract_pdfminer_chunk(
...
+        try:
+            wait_for_grobid(grobid_url, logger=active_logger)
+        except GrobidError as exc:
+            grobid_available = False
+            active_logger.warning(
+                "GROBID unavailable; using fallback extractor '%s' for all chunks: %s",
+                fallback_extractor,
+                exc,
+            )
+
```
[the remark this answers is not in the document]
Comment (Frank-Yong) [reply docdown/stages/extract.py:260]: I’ve added the availability gating and both new edge-case tests; I’ll run test_extract.py now to verify the behavior and ensure no regressions.

Ran terminal command: $pyRoot='C:\Users\email\AppData\Local\Programs\Python\Python314'; $scriptsDir=Join-Path $pyRoot 'Scripts'; $env:PATH = "$pyRoot;$scriptsDir;" + $env:PATH; pytest tests/test_extract.py -q

Implemented this review fix.

### What changed in extract.py
- GROBID readiness is now checked once whenever **either** configured extractor uses GROBID:
  - `extractor == "grobid"` **or** `fallback_extractor == "grobid"`.
- If GROBID is unavailable:
  - Per-chunk GROBID attempts are skipped.
  - If both primary and fallback are GROBID, each chunk gets a clear failed result:
    - `"GROBID unavailable and no non-GROBID extractor configured: ..."`
  - If fallback is GROBID and primary fails, fallback is skipped with clear failed result:
    - `"GROBID unavailable for fallback extraction: ..."`

This avoids repeated per-chunk connection timeouts while preserving chunk-level isolation.

### Tests added in test_extract.py
- `test_orchestrate_extraction_when_grobid_down_and_used_as_fallback_skips_calls`
- `test_orchestrate_extraction_when_both_extractors_are_grobid_and_down_fails_without_calls`

Both assert `extract_grobid_chunk` is never called when service is known down.

### Validation
- Ran `pytest test_extract.py -q`
- Result: `26 passed, 1 skipped` (plus existing `RequestsDependencyWarning`)

File: docdown/stages/extract.py [hunk 3]
```diff
@@ -240,3 +367,65 @@ def _resolve_logger(
...
+    return int(match.group(1))
+
+
+def _run_single_extractor(
+    extractor_name: str,
+    chunk_path: Path,
+    extracted_dir: Path,
+    chunk_number: int,
+    grobid_url: str,
+    logger: logging.Logger,
```
[the remark this answers is not in the document]
Comment (Frank-Yong) [reply docdown/stages/extract.py]: Ran terminal command: $pyRoot='C:\Users\email\AppData\Local\Programs\Python\Python314'; $scriptsDir=Join-Path $pyRoot 'Scripts'; $env:PATH = "$pyRoot;$scriptsDir;" + $env:PATH; pytest tests/test_extract.py -q

Applied.

I widened logger typing in extract.py to match actual adapter usage:

- Added alias:
  - `LogLike = logging.Logger | logging.LoggerAdapter`
- Updated `logger` annotations in:
  - `wait_for_grobid(...)`
  - `extract_grobid_chunk(...)`
  - `extract_pdfminer_chunk(...)`
  - `orchestrate_extraction(...)`
  - `_resolve_logger(...)`
  - `_run_single_extractor(...)`

This keeps type hints aligned with runtime behavior where chunk-aware adapters are used.

Validation:
- `pytest test_extract.py -q` -> `26 passed, 1 skipped` (plus existing `RequestsDependencyWarning`)

File: docdown/stages/extract.py [hunk 4]
```diff
@@ -18,6 +21,10 @@
 _DEFAULT_503_RETRIES = 3
 _DEFAULT_503_BACKOFF_BASE_SECONDS = 5
 _MAX_ERROR_BODY_EXCERPT = 300
+_VALID_EXTRACTORS = {"grobid", "pdfminer"}
+_CHUNK_STEM_PATTERN = re.compile(r"^chunk-(\d+)$")
```
[the remark this answers is not in the document]
Comment (Frank-Yong) [reply docdown/stages/extract.py]: I’ve tightened parsing and added coverage for `chunk-1.pdf`; now I’m running extraction tests to confirm everything still passes.

Ran terminal command: $pyRoot='C:\Users\email\AppData\Local\Programs\Python\Python314'; $scriptsDir=Join-Path $pyRoot 'Scripts'; $env:PATH = "$pyRoot;$scriptsDir;" + $env:PATH; pytest tests/test_extract.py -q

Implemented. I made chunk filename parsing strict to 4-digit naming and added a regression test.

### Changes
- Updated extract.py:
  - `_CHUNK_STEM_PATTERN` from `^chunk-(\d+)$` to `^chunk-(\d{4})$`
- Added test in test_extract.py:
  - `test_orchestrate_extraction_rejects_non_padded_chunk_number`
  - Confirms `chunk-1.pdf` is treated as invalid while a following valid chunk still processes.

This now matches the rest of the codebase’s `chunk-0001` convention and avoids confusing `chunk-1 -> chunk-0001` remapping behavior.

### Validation
- `pytest test_extract.py -q` -> `27 passed, 1 skipped` (plus existing `RequestsDependencyWarning`)

File: docdown/stages/extract.py [hunk 5]
```diff
@@ -201,6 +222,156 @@ def extract_pdfminer_chunk(
...
+        except ValueError as exc:
+            active_logger.error("Invalid chunk filename for extraction: %s: %s", chunk.name, exc)
+            results.append(
+                ExtractionResult(
+                    chunk_number=index,
+                    success=False,
+                    extractor=None,
+                    output_path=None,
+                    error=str(exc),
+                )
```
[the remark this answers is not in the document]
Comment (Frank-Yong) [reply docdown/stages/extract.py:276]: I’ve applied the sentinel contract and test assertions; I’m running the extraction tests now to verify this change is clean.

Ran terminal command: $pyRoot='C:\Users\email\AppData\Local\Programs\Python\Python314'; $scriptsDir=Join-Path $pyRoot 'Scripts'; $env:PATH = "$pyRoot;$scriptsDir;" + $env:PATH; pytest tests/test_extract.py -q

Implemented this review feedback with a clear sentinel.

### Change made
In extract.py, invalid chunk filename failures now produce:

- `ExtractionResult(chunk_number=0, success=False, extractor=None, output_path=None, error=...)`

instead of using loop index as a pseudo chunk number.

This avoids ambiguity between:
- real parsed chunk IDs (`1..N` from `chunk-NNNN.pdf`)
- parse failures (now always `0`)

### Tests updated
In test_extract.py, I added explicit assertions in both invalid-name orchestration tests:

- `results[0].chunk_number == 0`

### Validation
- `pytest test_extract.py -q` -> `27 passed, 1 skipped` (plus existing `RequestsDependencyWarning`)

File: docdown/stages/extract.py [hunk 6]
```diff
@@ -201,6 +222,156 @@ def extract_pdfminer_chunk(
...
+            grobid_unavailable_reason = str(exc)
+            active_logger.warning(
+                "GROBID unavailable; using fallback extractor '%s' for all chunks: %s",
+                fallback_extractor,
+                exc,
+            )
+
+    results: list[ExtractionResult] = []
+
+    for index, chunk_path in enumerate(chunk_paths, start=1):
```
[the remark this answers is not in the document]
Comment (Frank-Yong) [reply docdown/stages/extract.py]: Ran terminal command: $pyRoot='C:\Users\email\AppData\Local\Programs\Python\Python314'; $scriptsDir=Join-Path $pyRoot 'Scripts'; $env:PATH = "$pyRoot;$scriptsDir;" + $env:PATH; pytest tests/test_extract.py -q

Done. I removed the unused loop index in extract.py:

- From: `for index, chunk_path in enumerate(chunk_paths, start=1):`
- To: `for chunk_path in chunk_paths:`

Validation:
- `pytest test_extract.py -q` -> `27 passed, 1 skipped` (with the same existing `RequestsDependencyWarning`)

File: docdown/stages/extract.py [hunk 7]
```diff
@@ -232,11 +403,73 @@ def _body_excerpt(text: str, max_chars: int = _MAX_ERROR_BODY_EXCERPT) -> str:
...
+def _validate_extractor_name(value: str, field_name: str) -> None:
+    if value not in _VALID_EXTRACTORS:
+        raise ValueError(f"{field_name} must be one of: {sorted(_VALID_EXTRACTORS)}")
+
+
+def _chunk_number_from_path(path: Path) -> int:
+    match = _CHUNK_STEM_PATTERN.match(path.stem)
+    if not match:
+        raise ValueError(f"Chunk filename must match 'chunk-NNNN.pdf', got: {path.name}")
+    return int(match.group(1))
```
[the remark this answers is not in the document]
Comment (Frank-Yong) [reply docdown/stages/extract.py]: Ran terminal command: $pyRoot='C:\Users\email\AppData\Local\Programs\Python\Python314'; $scriptsDir=Join-Path $pyRoot 'Scripts'; $env:PATH = "$pyRoot;$scriptsDir;" + $env:PATH; pytest tests/test_extract.py -q

Implemented.

### Change in extract.py
`_chunk_number_from_path` now rejects `chunk-0000.pdf`:

- Parses 4-digit number.
- Validates `chunk_number >= 1`.
- Raises clear `ValueError` if zero:
  - `"Chunk filename must use chunk number >= 0001, got: ..."`

This keeps parsed chunk numbers aligned with 1-based conventions and preserves `chunk_number=0` as the orchestration sentinel for parse failures.

### Test added
In test_extract.py:
- `test_orchestrate_extraction_rejects_chunk_zero_filename`
  - Confirms `chunk-0000.pdf` yields failed result with sentinel `chunk_number == 0`
  - Confirms subsequent valid chunk still succeeds.

### Validation
- `pytest test_extract.py -q` -> `28 passed, 1 skipped` (plus existing `RequestsDependencyWarning`)

File: docdown/stages/extract.py [hunk 8]
```diff
@@ -201,6 +222,156 @@ def extract_pdfminer_chunk(
...
+    grobid_available = True
+    grobid_unavailable_reason: str | None = None
+    if grobid_required:
+        try:
+            wait_for_grobid(grobid_url, logger=active_logger)
+        except GrobidError as exc:
+            grobid_available = False
+            grobid_unavailable_reason = str(exc)
+            active_logger.warning(
+                "GROBID unavailable; using fallback extractor '%s' for all chunks: %s",
```
[the remark this answers is not in the document]
Comment (Frank-Yong) [reply docdown/stages/extract.py]: Ran terminal command: $pyRoot='C:\Users\email\AppData\Local\Programs\Python\Python314'; $scriptsDir=Join-Path $pyRoot 'Scripts'; $env:PATH = "$pyRoot;$scriptsDir;" + $env:PATH; pytest tests/test_extract.py -q

Updated the warning text in extract.py to match actual behavior.

### Change
- From:
  - `"GROBID unavailable; using fallback extractor '%s' for all chunks: %s"`
- To:
  - `"GROBID unavailable; skipping all grobid extraction attempts (extractor=%s, fallback=%s): %s"`

This is now accurate for all configurations, including when fallback is also `grobid`.

### Validation
- `pytest test_extract.py -q` -> `28 passed, 1 skipped` (with existing `RequestsDependencyWarning`)

File: docdown/stages/extract.py [hunk 9]
```diff
@@ -232,11 +403,76 @@ def _body_excerpt(text: str, max_chars: int = _MAX_ERROR_BODY_EXCERPT) -> str:
...
+            ExtractionResult(
+                chunk_number=chunk_number,
+                success=True,
+                extractor=ExtractorUsed.PDFMINER,
+                output_path=final_path,
+                error=None,
+            ),
+            None,
+        )
+    except (GrobidError, PdfMinerError, ValueError) as exc:
```
[the remark this answers is not in the document]
Comment (Frank-Yong) [reply docdown/stages/extract.py]: I’ve added the broader exception handling and a focused continuation test; now I’m running extraction tests to verify the failure-isolation contract still holds.

Ran terminal command: $pyRoot='C:\Users\email\AppData\Local\Programs\Python\Python314'; $scriptsDir=Join-Path $pyRoot 'Scripts'; $env:PATH = "$pyRoot;$scriptsDir;" + $env:PATH; pytest tests/test_extract.py -q

Implemented this robustness fix.

### What changed
In extract.py:

- `_run_single_extractor(...)` now also catches:
  - `OSError`
  - `RuntimeError`
- Existing catches (`GrobidError`, `PdfMinerError`, `ValueError`) remain.
- Result: chunk-level I/O/runtime failures are returned as per-chunk errors instead of aborting orchestration.

### Test added
In test_extract.py:

- `test_orchestrate_extraction_continues_after_primary_oserror`
  - Simulates `extract_grobid_chunk` raising `OSError` for `chunk-0001`
  - Confirms fallback succeeds for that chunk
  - Confirms `chunk-0002` still processes normally
  - Verifies orchestration continues across chunks

### Validation
- `pytest test_extract.py -q` -> `29 passed, 1 skipped` (with existing `RequestsDependencyWarning`)

File: docdown/stages/extract.py [hunk 10]
```diff
@@ -232,11 +404,76 @@ def _body_excerpt(text: str, max_chars: int = _MAX_ERROR_BODY_EXCERPT) -> str:
...
+def _validate_extractor_name(value: str, field_name: str) -> None:
+    if value not in _VALID_EXTRACTORS:
+        raise ValueError(f"{field_name} must be one of: {sorted(_VALID_EXTRACTORS)}")
+
+
+def _chunk_number_from_path(path: Path) -> int:
+    match = _CHUNK_STEM_PATTERN.match(path.stem)
+    if not match:
+        raise ValueError(f"Chunk filename must match 'chunk-NNNN.pdf', got: {path.name}")
+    chunk_number = int(match.group(1))
```
[the remark this answers is not in the document]
Comment (Frank-Yong) [reply docdown/stages/extract.py:430]: I’ve added both the extension guard and regression coverage; next I’m running extraction tests to confirm the contract change is stable.

Ran terminal command: $pyRoot='C:\Users\email\AppData\Local\Programs\Python\Python314'; $scriptsDir=Join-Path $pyRoot 'Scripts'; $env:PATH = "$pyRoot;$scriptsDir;" + $env:PATH; pytest tests/test_extract.py -q

Implemented this review fix.

### Changes in extract.py
- `_chunk_number_from_path` now enforces PDF extension first:
  - `if path.suffix.lower() != ".pdf": raise ValueError(...)`

This aligns behavior with the function contract/error messaging and prevents files like `chunk-0001.txt` from being treated as valid chunk PDFs.

### Test added
In test_extract.py:
- `test_orchestrate_extraction_rejects_non_pdf_chunk_extension`
  - Confirms non-PDF chunk filename is marked failed with sentinel `chunk_number == 0`
  - Confirms next valid `.pdf` chunk still succeeds (no orchestration abort)

### Validation
- `pytest test_extract.py -q` -> `30 passed, 1 skipped` (plus existing `RequestsDependencyWarning`)

File: docdown/stages/extract.py [hunk 11]
```diff
@@ -18,6 +21,10 @@
 _DEFAULT_503_RETRIES = 3
 _DEFAULT_503_BACKOFF_BASE_SECONDS = 5
 _MAX_ERROR_BODY_EXCERPT = 300
+_VALID_EXTRACTORS = {"grobid", "pdfminer"}
+_CHUNK_STEM_PATTERN = re.compile(r"^chunk-(\d{4})$")
```
[the remark this answers is not in the document]
Comment (Frank-Yong) [reply docdown/stages/extract.py]: Ran terminal command: $pyRoot='C:\Users\email\AppData\Local\Programs\Python\Python314'; $scriptsDir=Join-Path $pyRoot 'Scripts'; $env:PATH = "$pyRoot;$scriptsDir;" + $env:PATH; pytest tests/test_extract.py -q

Implemented.

### Change
In extract.py:
- Removed hardcoded `_VALID_EXTRACTORS = {"grobid", "pdfminer"}`
- Added enum-derived set right after `ExtractorUsed`:
  - `_VALID_EXTRACTORS = {member.value for member in ExtractorUsed}`

This keeps validation in sync automatically if new extractors are added to the enum.

### Validation
- `pytest test_extract.py -q` -> `30 passed, 1 skipped` (with existing `RequestsDependencyWarning`)

File: docs/plan/task-3.3-extraction-orchestration.md [hunk 12]
```diff
@@ -11,14 +11,18 @@ Implement the orchestration layer that runs the primary extractor, falls back wh
...
-- [ ] Extraction results are returned as a list for downstream stages to consume.
-- [ ] Summary logged: N succeeded (grobid), M succeeded (pdfminer), K failed.
-- [ ] Unit tests cover: all-succeed, partial-failure, full-GROBID-down, all-fail scenarios.
+- [x] For each chunk: try primary extractor → on failure, try fallback → on failure, mark as failed.
+- [x] Extractor choice is driven by config (`extractor`, `fallback_extractor`).
+- [x] Per-chunk result is tracked: `success (grobid)`, `success (pdfminer)`, or `failed`.
+- [x] Failed chunks do not block processing of other chunks.
+- [x] If GROBID is entirely unreachable, all chunks fall back to pdfminer without per-chunk retries.
+- [x] Extraction results are returned as a list for downstream stages to consume.
+- [x] Summary logged: N succeeded (grobid), M succeeded (pdfminer), K failed.
```
[the remark this answers is not in the document]
Comment (Frank-Yong) [reply docs/plan/task-3.3-extraction-orchestration.md:20]: Updated that acceptance criterion in `docs/plan/task-3.3-extraction-orchestration.md:17` to match actual behavior.

New wording:
- `If GROBID is entirely unreachable, orchestration skips GROBID attempts and uses the configured fallback extractor without per-chunk GROBID retries (or records clear failures when fallback is also GROBID).`

This now aligns with the implementation/config surface and avoids implying hardcoded pdfminer fallback.

File: docdown/stages/extract.py [hunk 13]
```diff
@@ -232,11 +404,76 @@ def _body_excerpt(text: str, max_chars: int = _MAX_ERROR_BODY_EXCERPT) -> str:
...
-    logger: logging.Logger | None,
+    logger: LogLike | None,
     chunk_number: int | None,
-) -> logging.Logger | ChunkAdapter:
+) -> LogLike | ChunkAdapter:
     if chunk_number is None:
         return logger or get_logger()
     if logger is None:
         return get_chunk_logger(chunk_number)
     return ChunkAdapter(logger, {"chunk": chunk_number})
```
[the remark this answers is not in the document]
Comment (Frank-Yong) [reply docdown/stages/extract.py:416]: Ran terminal command: $pyRoot='C:\Users\email\AppData\Local\Programs\Python\Python314'; $scriptsDir=Join-Path $pyRoot 'Scripts'; $env:PATH = "$pyRoot;$scriptsDir;" + $env:PATH; pytest tests/test_extract.py -q

Fixed. `_resolve_logger` now returns just `LogLike` in extract.py, which is accurate since `ChunkAdapter` is already a `LoggerAdapter` and covered by `LogLike`.

Validation:
- `pytest test_extract.py -q` -> `30 passed, 1 skipped` (plus existing `RequestsDependencyWarning`)
